librz/analysis/rzil/rzil_trace.c

```diff
@@ -0,0 +1,237 @@
+		return;
+	}
+
+	RzILTraceMemOp *mem_write = RZ_NEW0(RzILTraceMemOp);
+	if (!mem_write) {
+		RZ_LOG_ERROR("RZIL cannot init memory write trace\n");
+		return;
+	}
+	mem_write->behavior = RZ_IL_TRACE_OP_WRITE;
+	mem_write->addr = rz_il_bv_to_ut64(addr);
+	bv_to_databuf(mem_write->data_buf, data);
+
+	rzil_add_mem_trace(rzil->trace, mem_write);
+}
+
+static void rz_analysis_rzil_trace_focus_reg_read(RzAnalysis *analysis, RZ_NONNULL RzAnalysisRzil *rzil, RZ_NONNULL RzILOp single_op) {
+	rz_return_if_fail(rzil);
+	rz_return_if_fail(single_op);
+	RzILOpVar op_var = single_op->op.var;
+	RzILVM vm = rzil->vm;
+
+	const char *reg_name = rz_str_constpool_get(&analysis->constpool, op_var->v);
+	ut64 data = rz_il_bv_to_ut64(rz_il_get_bv_temp(vm, op_var->ret));
+
+	RzILTraceRegOp *reg_read = RZ_NEW0(RzILTraceRegOp);
+	if (!reg_read) {
+		RZ_LOG_ERROR("RZIL cannot init register read trace\n");
+		return;
+	}
+	reg_read->behavior = RZ_IL_TRACE_OP_READ;
+	reg_read->reg_name = reg_name;
+	reg_read->value = data;
+
+	rzil_add_reg_trace(rzil->trace, reg_read);
+}
+
+static void rz_analysis_rzil_trace_focus_reg_write(RzAnalysis *analysis, RZ_NONNULL RzAnalysisRzil *rzil, RZ_NONNULL RzILOp single_op) {
+	rz_return_if_fail(rzil);
+	rz_return_if_fail(single_op);
+	RzILOpSet op_set = single_op->op.set;
+	RzILVM vm = rzil->vm;
+
+	const char *reg_name = rz_str_constpool_get(&analysis->constpool, op_set->v);
+	ut64 data = rz_il_bv_to_ut64(rz_il_get_bv_temp(vm, op_set->x));
+
+	RzILTraceRegOp *reg_write = RZ_NEW0(RzILTraceRegOp);
+	if (!reg_write) {
+		RZ_LOG_ERROR("RZIL cannot init register write trace\n");
+		return;
+	}
+	reg_write->behavior = RZ_IL_TRACE_OP_WRITE;
+	reg_write->reg_name = reg_name;
+	reg_write->value = data;
+
+	rzil_add_reg_trace(rzil->trace, reg_write);
+}
+
+static void rz_analysis_rzil_trace_focus(RzAnalysis *analysis, RZ_NONNULL RzAnalysisRzil *rzil, RZ_NONNULL RzILOp single_op) {
+	// focus those op only
+	switch (single_op->code) {
+	case RZIL_OP_LOAD:
+		rz_analysis_rzil_trace_focus_mem_read(analysis, rzil, single_op);
+		break;
+	case RZIL_OP_STORE:
+		rz_analysis_rzil_trace_focus_mem_write(analysis, rzil, single_op);
+		break;
+	case RZIL_OP_SET:
+		rz_analysis_rzil_trace_focus_reg_write(analysis, rzil, single_op);
+		break;
+	case RZIL_OP_VAR:
+		rz_analysis_rzil_trace_focus_reg_read(analysis, rzil, single_op);
+		break;
+	default:
+		// don't need to trace info
+		break;
+	}
+}
+
+/**
+ * This function should be called after executing the RZIL op
+ * Collect trace info (target and data of mem/reg read/write)
+ * \param analysis RzAnalysis
+ * \param rzil RZIL instance
+ * \param op RzAnalysisRzilOp, a general RZIL op structure (Designed for switching between different implementations of RZIL op struct)
+ */
+RZ_API void rz_analysis_rzil_trace_op(RzAnalysis *analysis, RZ_NONNULL RzAnalysisRzil *rzil, RZ_NONNULL RzAnalysisRzilOp *op) {
+	// TODO : rewrite this file when migrate to new op structure
```

Comment:
"file" or "function"? If file, then commend should be moved to the top, if function it should be named properly.

Reply:
"Functions in the file", I will move it to the top